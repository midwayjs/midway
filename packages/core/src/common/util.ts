import { dirname, resolve, sep, extname } from 'path';

export const safeRequire = p => {
  if (p.startsWith(`.${sep}`) || p.startsWith(`..${sep}`)) {
    p = resolve(dirname(module.parent.filename), p);
  }
  try {
    return require(p);
  } catch (err) {
    return undefined;
  }
};

export const isPath = (p): boolean => {
  // eslint-disable-next-line no-useless-escape
  if (/(^[\.\/])|:|\\/.test(p)) {
    return true;
  }
  return false;
};

/**
 *  safelyGet(['a','b'],{a: {b: 2}})  // => 2
 *  safelyGet(['a','b'],{c: {b: 2}})  // => undefined
 *  safelyGet(['a','1'],{a: {"1": 2}})  // => 2
 *  safelyGet(['a','1'],{a: {b: 2}})  // => undefined
 *  safelyGet('a.b',{a: {b: 2}})  // => 2
 *  safelyGet('a.b',{c: {b: 2}})  // => undefined
 */
export function safelyGet(list: string | string[], obj?: object): any {
  if (arguments.length === 1) {
    return (_obj: object) => safelyGet(list, _obj);
  }

  if (typeof obj === 'undefined' || typeof obj !== 'object' || obj === null) {
    return void 0;
  }
  const pathArrValue = typeof list === 'string' ? list.split('.') : list;
  let willReturn: any = obj;

  for (const key of pathArrValue) {
    if (typeof willReturn === 'undefined' || willReturn === null) {
      return void 0;
    } else if (typeof willReturn !== 'object') {
      return void 0;
    }
    willReturn = willReturn[key];
  }

  return willReturn;
}

/**
 * 剔除 @ 符号
 * @param provideId provideId
 */
export function parsePrefix(provideId: string) {
  if (provideId.includes('@')) {
    return provideId.substr(1);
  }
  return provideId;
}

export function isPathEqual(one: string, two: string) {
  if (!one || !two) {
    return false;
  }
  const ext = extname(one);
  return one.replace(ext, '') === two;
}
