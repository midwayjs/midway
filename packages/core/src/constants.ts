/**
 * 静态参数
 *
 */
export const KEYS = {
  OBJECTS_ELEMENT: 'objects',
  OBJECT_ELEMENT: 'object',
  IMPORT_ELEMENT: 'import',
  PROPERTY_ELEMENT: 'property',
  LIST_ELEMENT: 'list',
  MAP_ELEMENT: 'map',
  ENTRY_ELEMENT: 'entry',
  VALUE_ELEMENT: 'value',
  PROPS_ELEMENT: 'props',
  PROP_ELEMENT: 'prop',
  SET_ELEMENT: 'set',
  CONSTRUCTOR_ARG_ELEMENT: 'constructor-arg',
  REF_ELEMENT: 'ref',
  JSON_ELEMENT: 'json',
  CONFIGURATION_ELEMENT: 'configuration',

  ID_ATTRIBUTE: 'id',
  PATH_ATTRIBUTE: 'path',
  DIRECT_ATTRIBUTE: 'direct',
  AUTOWIRE_ATTRIBUTE: 'autowire',
  ASYNC_ATTRIBUTE: 'async',
  NAME_ATTRIBUTE: 'name',
  REF_ATTRIBUTE: 'ref',
  KEY_ATTRIBUTE: 'key',
  VALUE_ATTRIBUTE: 'value',
  TYPE_ATTRIBUTE: 'type',
  EXTERNAL_ATTRIBUTE: 'external',
  OBJECT_ATTRIBUTE: 'object',
  RESOURCE_ATTRIBUTE: 'resource',
  SCOPE_ATTRIBUTE: 'scope',

  ASPECT_ELEMENT: 'aspect',
  AROUND_ELEMENT: 'around',
  EXPRESSION_ATTRIBUTE: 'expression',
  EXECUTE_ATTRIBUTE: 'execute',
};

export const FUNCTION_INJECT_KEY = 'midway:function_inject_key';
export const MIDWAY_LOGGER_WRITEABLE_DIR = 'MIDWAY_LOGGER_WRITEABLE_DIR';
export const REQUEST_CTX_KEY = 'ctx';
export const REQUEST_OBJ_CTX_KEY = '_req_ctx';
export const HTTP_SERVER_KEY = '_midway_http_server';
export const REQUEST_CTX_LOGGER_CACHE_KEY = '_midway_ctx_logger_cache';

export const ASYNC_CONTEXT_KEY = Symbol('ASYNC_CONTEXT_KEY');
export const ASYNC_CONTEXT_MANAGER_KEY = 'MIDWAY_ASYNC_CONTEXT_MANAGER_KEY';

export const DEFAULT_PATTERN = [
  '**/**.ts',
  '**/**.js',
  '**/**.mts',
  '**/**.mjs',
  '**/**.cts',
  '**/**.cjs',
];

export const IGNORE_PATTERN = ['**/**.d.ts', '**/**.d.mts', '**/**.d.cts'];
