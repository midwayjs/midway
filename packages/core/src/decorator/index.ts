// common
export * from './common/provide';
export * from './common/inject';
export * from './common/pipeline';
export * from './common/aspect';
export * from './common/autoload';
export * from './common/configuration';
export * from './common/objectDef';
export * from './common/framework';
export * from './common/filter';
export * from './common/middleware';
export * from './common/guard';
export * from './common/pipe';
export * from './common/mock';

// faas
export * from './faas/serverlessTrigger';

// web
export * from './web/controller';
export * from './web/paramMapping';
export * from './web/requestMapping';
export * from './web/response';

// other
export * from './constant';
export * from './decoratorManager';

// microservice
export * from './microservice/consumer';
export * from './microservice/provider';
export * from './microservice/rabbitmqListener';
export * from './microservice/kafkaListener';

// rpc
export * from './rpc/hsf';

// task
export * from './task/queue';
export * from './task/task';
export * from './task/taskLocal';
export * from './task/schedule';

// ws
export * from './ws/webSocketController';
export * from './ws/webSocketEvent';
