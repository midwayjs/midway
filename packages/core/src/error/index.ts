export * from './base';
export { HttpStatus, httpError } from './http';
export * from './framework';
