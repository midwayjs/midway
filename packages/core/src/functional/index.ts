export { defineConfiguration } from './configuration';
export * from './hooks';
export * from './interface';
