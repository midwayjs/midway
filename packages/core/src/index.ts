export {
  ObjectIdentifier,
  ObjectDefinitionOptions,
  IManagedInstance,
  ScopeEnum,
  saveClassMetadata,
  attachClassMetadata,
  getClassMetadata,
  saveMethodDataToClass,
  attachMethodDataToClass,
  getMethodDataFromClass,
  listMethodDataFromClass,
  saveMethodMetadata,
  attachMethodMetadata,
  getMethodMetadata,
  savePropertyDataToClass,
  attachPropertyDataToClass,
  getPropertyDataFromClass,
  listPropertyDataFromClass,
  savePropertyMetadata,
  attachPropertyMetadata,
  getPropertyMetadata,
  savePreloadModule,
  listPreloadModule,
  saveModule,
  listModule,
  resetModule,
  clearAllModule,
  getParamNames,
  getProviderId,
  getObjectDefinition,
  classNamed,
  generateProvideId
} from '@midwayjs/decorator';
export * from './interface';
export { ContainerLoader } from './loader';
export { MidwayContainer } from './context/midwayContainer';
export { MidwayRequestContainer } from './context/requestContainer';
export { BaseFramework } from './baseFramework';
export * from './context/providerWrapper';
export * from './common/constants';
export {
  safelyGet,
  safeRequire,
  parsePrefix,
} from './common/util';
export * from './features';
export * from './util/webRouterParam';
export { plainToClass, classToPlain } from 'class-transformer';
