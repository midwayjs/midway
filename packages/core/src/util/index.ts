import { dirname, resolve, sep, posix } from 'path';
import { readFileSync } from 'fs';
import { debuglog } from 'util';

const debug = debuglog('midway:container:util');

export const isDevelopmentEnvironment = env => {
  return ['local', 'test', 'unittest'].includes(env);
};

export const getCurrentEnvironment = () => {
  return process.env['MIDWAY_SERVER_ENV'] || process.env['NODE_ENV'] || 'prod';
};

export const safeRequire = (p, enabledCache = true) => {
  if (p.startsWith(`.${sep}`) || p.startsWith(`..${sep}`)) {
    p = resolve(dirname(module.parent.filename), p);
  }
  try {
    if (enabledCache) {
      return require(p);
    } else {
      const content = readFileSync(p, {
        encoding: 'utf-8',
      });
      return JSON.parse(content);
    }
  } catch (err) {
    debug('SafeRequire Warning', err.message);
    return undefined;
  }
};

/**
 *  safelyGet(['a','b'],{a: {b: 2}})  // => 2
 *  safelyGet(['a','b'],{c: {b: 2}})  // => undefined
 *  safelyGet(['a','1'],{a: {"1": 2}})  // => 2
 *  safelyGet(['a','1'],{a: {b: 2}})  // => undefined
 *  safelyGet('a.b',{a: {b: 2}})  // => 2
 *  safelyGet('a.b',{c: {b: 2}})  // => undefined
 */
export function safelyGet(
  list: string | string[],
  obj?: Record<string, unknown>
): any {
  if (arguments.length === 1) {
    return (_obj: Record<string, unknown>) => safelyGet(list, _obj);
  }

  if (typeof obj === 'undefined' || typeof obj !== 'object' || obj === null) {
    return void 0;
  }
  const pathArrValue = typeof list === 'string' ? list.split('.') : list;
  let willReturn: any = obj;

  for (const key of pathArrValue) {
    if (typeof willReturn === 'undefined' || willReturn === null) {
      return void 0;
    } else if (typeof willReturn !== 'object') {
      return void 0;
    }
    willReturn = willReturn[key];
  }

  return willReturn;
}

/**
 * 剔除 @ 符号
 * @param provideId provideId
 */
export function parsePrefix(provideId: string) {
  if (provideId.includes('@')) {
    return provideId.substr(1);
  }
  return provideId;
}

export function getUserHome() {
  return process.env[process.platform === 'win32' ? 'USERPROFILE' : 'HOME'];
}

export function joinURLPath(...strArray) {
  if (strArray.length === 0) {
    return '';
  }
  let p = posix.join(...strArray);
  p = p.replace(/\/+$/, '');
  if (!/^\//.test(p)) {
    p = '/' + p;
  }
  return p;
}

/**
 * 代理目标所有的原型方法，不包括构造器和内部隐藏方法
 * @param derivedCtor
 * @param constructors
 */
export function delegateTargetPrototypeMethod(
  derivedCtor: any,
  constructors: any[]
) {
  constructors.forEach(baseCtor => {
    Object.getOwnPropertyNames(baseCtor.prototype).forEach(name => {
      if (name !== 'constructor' && !/^_/.test(name)) {
        derivedCtor.prototype[name] = function (...args) {
          return this.instance[name](...args);
        };
      }
    });
  });
}

/**
 * 代理目标原型上的特定方法
 * @param derivedCtor
 * @param methods
 */
export function delegateTargetMethod(derivedCtor: any, methods: string[]) {
  methods.forEach(name => {
    derivedCtor.prototype[name] = function (...args) {
      return this.instance[name](...args);
    };
  });
}

/**
 * 代理目标原型属性
 * @param derivedCtor
 * @param properties
 */
export function delegateTargetProperties(
  derivedCtor: any,
  properties: string[]
) {
  properties.forEach(name => {
    Object.defineProperty(derivedCtor.prototype, name, {
      get() {
        return this.instance[name];
      },
    });
  });
}
