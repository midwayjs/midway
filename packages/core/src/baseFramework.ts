import {
  IConfigurationOptions,
  IMidwayApplication,
  IMidwayBootstrapOptions,
  IMidwayContainer,
  IMidwayFramework,
  MidwayFrameworkType,
  MidwayProcessTypeEnum,
} from './interface';
import { ContainerLoader } from './';
import { APPLICATION_KEY } from '@midwayjs/decorator';

export abstract class BaseFramework<
  APP extends IMidwayApplication,
  T extends IConfigurationOptions
> implements IMidwayFramework<APP, T> {
  protected isTsMode = true;
  protected baseDir: string;
  protected appDir: string;
  protected containerLoader: ContainerLoader;
  public configurationOptions: T;
  public app: APP;

  public configure(options: T): BaseFramework<APP, T> {
    this.configurationOptions = options;
    return this;
  }

  public async initialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {
    this.baseDir = options.baseDir;
    this.appDir = options.appDir;

    this.containerLoader = new ContainerLoader({
      baseDir: this.baseDir,
      isTsMode: this.isTsMode,
      preloadModules: options.preloadModules || [],
    });

    /**
     * initialize containerLoader and initialize ioc container instance
     */
    await this.beforeInitialize(options);
    this.containerLoader.initialize();

    /**
     * load directory and bind files to ioc container
     */
    await this.beforeDirectoryLoad(options);
    const applicationContext = this.containerLoader.getApplicationContext();
    applicationContext.registerObject('baseDir', this.baseDir);
    applicationContext.registerObject('appDir', this.appDir);
    // 如果没有关闭autoLoad 则进行load
    this.containerLoader.loadDirectory(options);

    // register app
    this.containerLoader.registerHook(APPLICATION_KEY, () => {
      return this.getApplication();
    });

    await this.afterDirectoryLoad(options);

    if (!this.app.getApplicationContext) {
      this.app = this.defineApplicationProperties(this.app);
    }

    /**
     * start to load configuration and lifeCycle
     */
    await this.containerLoader.refresh();
    await this.afterInitialize(options);
  }

  public getApplicationContext(): IMidwayContainer {
    return this.containerLoader.getApplicationContext();
  }

  public getConfiguration(key?: string): any {
    return this.getApplicationContext()
      .getConfigService()
      .getConfiguration(key);
  }

  public getCurrentEnvironment() {
    return this.getApplicationContext()
      .getEnvironmentService()
      .getCurrentEnvironment();
  }

  public abstract getFrameworkType(): MidwayFrameworkType;

  public abstract getApplication(): APP;

  public abstract run(): Promise<void>;

  public async stop(): Promise<void> {
    await this.beforeStop();
    await this.containerLoader.stop();
  }

  protected defineApplicationProperties(app: APP): APP {
    return Object.assign(app, {
      getBaseDir: () => {
        return this.baseDir;
      },

      getAppDir: () => {
        return this.appDir;
      },

      getEnv: () => {
        return this.getApplicationContext()
          .getEnvironmentService()
          .getCurrentEnvironment();
      },

      getConfig: (key?: string) => {
        return this.getApplicationContext()
          .getConfigService()
          .getConfiguration(key);
      },

      getFrameworkType: () => {
        return this.getFrameworkType();
      },

      getProcessType: () => {
        return MidwayProcessTypeEnum.APPLICATION;
      },
    });
  }

  protected async beforeStop(): Promise<void> {}

  protected async beforeInitialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected async beforeDirectoryLoad(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected async afterDirectoryLoad(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected abstract async afterInitialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void>;
}
