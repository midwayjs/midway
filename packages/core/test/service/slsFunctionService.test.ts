import { createLightFramework, matchObjectPropertyInArray } from '../util';
import * as path from 'path';
import { MidwayServerlessFunctionService, ServerlessTriggerType } from '../../src';

describe('/test/service/slsFunction.test.ts', function () {

  it('should test dynamic add function', async () => {
    const midwayServerlessFunctionService = new MidwayServerlessFunctionService();
    midwayServerlessFunctionService.addServerlessFunction(async (ctx, event) => {
      return 'hello world';
    }, {
      type: ServerlessTriggerType.HTTP,
      metadata: {
        method: 'get',
        path: '/api/hello'
      },
      functionName: 'hello',
      handlerName: 'index.hello',
    });

    midwayServerlessFunctionService.addServerlessFunction(async (ctx, event) => {
      return 'hello world';
    }, {
      type: ServerlessTriggerType.HSF,
      metadata: {
        name: 'customHSF'
      },
      functionName: 'hsf',
      handlerName: 'index.hsf',
    });

    const result = await midwayServerlessFunctionService.getFunctionList();
    expect(result).toMatchSnapshot();
    expect(typeof result[0].method).toEqual('function');
  });

  it('should test set correct function name', async () => {
    const framework = await createLightFramework(path.join(
      __dirname,
      '../fixtures/app-with-function-name/src'
    ));

    const midwayServerlessFunctionService = framework.getApplicationContext().get(MidwayServerlessFunctionService);

    midwayServerlessFunctionService.addServerlessFunction(async (ctx, event) => {
      return 'hello world';
    }, {
      type: ServerlessTriggerType.HTTP,
      metadata: {
        method: 'get',
        path: '/api/hello'
      },
      functionName: 'hello111',
      handlerName: 'index.hello',
    });

    const result = {
      'localTest.hello1': 'aaa1',
      'localTest.hello2': 'localTest-hello2',
      'localTest.hello3': 'aaa3',
      'localTest.hello4': 'aaa4',
      'index.hello': 'hello111'
    }

    const collector = await framework.getApplicationContext().getAsync(MidwayServerlessFunctionService);
    const fnList = await collector.getFunctionList();
    let size = 0;
    for(const fn of fnList) {
      if (result[fn.funcHandlerName] && fn.functionName === result[fn.funcHandlerName]) {
        size++;
      }
    }
    expect(size).toEqual(Object.keys(result).length);
  });

  it('should test with function router', async () => {
    const framework = await createLightFramework(path.join(
      __dirname,
      '../fixtures/base-app-func-router/src'
    ));
    const midwayWebRouterService = await framework.getApplicationContext().getAsync(MidwayServerlessFunctionService);
    let result = await midwayWebRouterService.getFlattenRouterTable();
    result = result.map(item => {
      delete item.id;
      return item;
    });
    expect(result).toMatchSnapshot();
  });

  it('should test with serverless trigger', async () => {
    const framework = await createLightFramework(path.join(
      __dirname,
      '../fixtures/app-with-serverless-trigger/src'
    ));
    const collector = await framework.getApplicationContext().getAsync(MidwayServerlessFunctionService);
    const result = await collector.getFlattenRouterTable();
    expect(matchObjectPropertyInArray(result, {
      functionName: 'helloAliyunService-handleTimerEvent',
      functionTriggerName: 'timer',
      functionTriggerMetadata: {
        type: 'cron',
        value: '0 0 4 * * *',
        name: 'custom_timer',
      },
      functionMetadata: {
        functionName: 'helloAliyunService-handleTimerEvent'
      }
    })).toBeTruthy();
    expect(matchObjectPropertyInArray(result, {
      functionName: 'hello_bbb',
      functionMetadata: {
        functionName: 'hello_bbb',
        concurrency: 2,
      },
    })).toBeTruthy();
  });

  it('should test duplicate router', async () => {
    const framework = await createLightFramework(path.join(
      __dirname,
      '../fixtures/app-with-duplicate-router/src'
    ));
    const collector = await framework.getApplicationContext().getAsync(MidwayServerlessFunctionService);
    await expect(collector.getFunctionList()).rejects.toThrow('Duplicate router')
  });

});
