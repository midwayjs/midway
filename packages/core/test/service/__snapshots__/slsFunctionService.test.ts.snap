// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/test/service/slsFunction.test.ts should test dynamic add function 1`] = `
[
  {
    "_category": 2,
    "_level": 2,
    "_paramString": "",
    "_pureRouter": "/api/hello",
    "_weight": 2,
    "controllerId": "",
    "controllerMiddleware": [],
    "description": "",
    "fullUrl": "/api/hello",
    "fullUrlCompiledRegexp": /\\^\\\\/api\\\\/hello\\[\\\\/#\\\\\\?\\]\\?\\$/i,
    "fullUrlFlattenString": "/api/hello",
    "funcHandlerName": "index.hello",
    "functionMetadata": {
      "functionName": "hello",
    },
    "functionName": "hello",
    "functionTriggerMetadata": {
      "method": "get",
      "path": "/api/hello",
    },
    "functionTriggerName": undefined,
    "handlerName": "",
    "id": null,
    "method": [Function],
    "middleware": [],
    "requestMetadata": [],
    "requestMethod": "get",
    "responseMetadata": [],
    "summary": "",
    "url": "/api/hello",
  },
  {
    "_category": 2,
    "_level": 0,
    "_paramString": "",
    "_pureRouter": "",
    "_weight": 0,
    "controllerId": "",
    "controllerMiddleware": [],
    "description": "",
    "funcHandlerName": "index.hsf",
    "functionMetadata": {
      "functionName": "hsf",
    },
    "functionName": "hsf",
    "functionTriggerMetadata": {
      "name": "customHSF",
    },
    "functionTriggerName": "customHSF",
    "handlerName": "",
    "id": null,
    "method": [Function],
    "middleware": [],
    "requestMetadata": [],
    "requestMethod": "",
    "responseMetadata": [],
    "summary": "",
    "url": "",
  },
]
`;

exports[`/test/service/slsFunction.test.ts should test with function router 1`] = `
[
  {
    "_category": 2,
    "_level": 1,
    "_paramString": "",
    "_pureRouter": "/",
    "_weight": 0,
    "controllerClz": [Function],
    "controllerId": "apiController",
    "controllerMiddleware": [
      "auth",
    ],
    "description": "",
    "fullUrl": "/api",
    "fullUrlFlattenString": "/api",
    "funcHandlerName": "apiController.homeSet",
    "handlerName": "apiController.homeSet",
    "method": "homeSet",
    "middleware": [
      "auth2",
    ],
    "prefix": "/api",
    "requestMetadata": [],
    "requestMethod": "get",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "/",
  },
  {
    "_category": 2,
    "_level": 1,
    "_paramString": "",
    "_pureRouter": "/invoke2",
    "_weight": 1,
    "controllerId": "helloHttpService",
    "controllerMiddleware": [],
    "description": "",
    "fullUrl": "/invoke2",
    "fullUrlFlattenString": "/invoke2",
    "funcHandlerName": "index.invoke2",
    "functionMetadata": {
      "functionName": "invoke2",
    },
    "functionName": "invoke2",
    "functionTriggerMetadata": {
      "functionName": "invoke2",
      "handlerName": "index.invoke2",
      "method": "get",
      "path": "/invoke2",
    },
    "functionTriggerName": "http",
    "handlerName": "helloHttpService.invoke2",
    "method": "invoke2",
    "middleware": [],
    "prefix": "/",
    "requestMetadata": [],
    "requestMethod": "get",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "/invoke2",
  },
  {
    "_category": 2,
    "_level": 1,
    "_paramString": "",
    "_pureRouter": "/upload",
    "_weight": 1,
    "controllerId": "helloHttpService",
    "controllerMiddleware": [],
    "description": "",
    "fullUrl": "/upload",
    "fullUrlFlattenString": "/upload",
    "funcHandlerName": "helloHttpService.upload",
    "functionMetadata": {
      "functionName": "helloHttpService-upload",
    },
    "functionName": "helloHttpService-upload",
    "functionTriggerMetadata": {
      "method": "get",
      "middleware": [
        "fmw:upload",
      ],
      "path": "/upload",
    },
    "functionTriggerName": "http",
    "handlerName": "helloHttpService.upload",
    "method": "upload",
    "middleware": [
      "fmw:upload",
    ],
    "prefix": "/",
    "requestMetadata": [],
    "requestMethod": "get",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "/upload",
  },
  {
    "_category": 2,
    "_level": 1,
    "_paramString": "",
    "_pureRouter": "/update",
    "_weight": 1,
    "controllerId": "helloHttpService",
    "controllerMiddleware": [],
    "description": "",
    "fullUrl": "/update",
    "fullUrlFlattenString": "/update",
    "funcHandlerName": "helloHttpService.invoke",
    "functionMetadata": {
      "functionName": "helloHttpService-invoke",
    },
    "functionName": "helloHttpService-invoke",
    "functionTriggerMetadata": {
      "method": "post",
      "middleware": [
        "auth",
      ],
      "path": "/update",
    },
    "functionTriggerName": "apigw",
    "handlerName": "helloHttpService.invoke",
    "method": "invoke",
    "middleware": [
      "auth",
    ],
    "prefix": "/",
    "requestMetadata": [],
    "requestMethod": "post",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "/update",
  },
  {
    "_category": 2,
    "_level": 1,
    "_paramString": "",
    "_pureRouter": "/invoke",
    "_weight": 1,
    "controllerId": "helloHttpService",
    "controllerMiddleware": [],
    "description": "",
    "fullUrl": "/invoke",
    "fullUrlFlattenString": "/invoke",
    "funcHandlerName": "helloHttpService.invoke",
    "functionMetadata": {
      "functionName": "helloHttpService-invoke",
    },
    "functionName": "helloHttpService-invoke",
    "functionTriggerMetadata": {
      "method": "get",
      "path": "/invoke",
    },
    "functionTriggerName": "http",
    "handlerName": "helloHttpService.invoke",
    "method": "invoke",
    "middleware": [],
    "prefix": "/",
    "requestMetadata": [],
    "requestMethod": "get",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "/invoke",
  },
  {
    "_category": 2,
    "_level": 0,
    "_paramString": "",
    "_pureRouter": "",
    "_weight": 0,
    "controllerId": "helloHttpService",
    "controllerMiddleware": [],
    "description": "",
    "funcHandlerName": "helloHttpService.upload",
    "functionMetadata": {
      "functionName": "helloHttpService-upload",
    },
    "functionName": "helloHttpService-upload",
    "functionTriggerMetadata": {},
    "functionTriggerName": "hsf",
    "handlerName": "helloHttpService.upload",
    "method": "upload",
    "middleware": [],
    "prefix": "/",
    "requestMetadata": [],
    "requestMethod": "",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "",
  },
  {
    "_category": 2,
    "_level": 0,
    "_paramString": "",
    "_pureRouter": "",
    "_weight": 0,
    "controllerId": "helloHttpService",
    "controllerMiddleware": [],
    "description": "",
    "funcHandlerName": "helloHttpService.invoke",
    "functionMetadata": {
      "functionName": "helloHttpService-invoke",
    },
    "functionName": "helloHttpService-invoke",
    "functionTriggerMetadata": {
      "payload": "",
      "type": "every",
      "value": "5m",
    },
    "functionTriggerName": "timer",
    "handlerName": "helloHttpService.invoke",
    "method": "invoke",
    "middleware": [],
    "prefix": "/",
    "requestMetadata": [],
    "requestMethod": "",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "",
  },
  {
    "_category": 2,
    "_level": 0,
    "_paramString": "",
    "_pureRouter": "",
    "_weight": 0,
    "controllerId": "helloHttpService",
    "controllerMiddleware": [],
    "description": "",
    "funcHandlerName": "index.invoke3",
    "functionMetadata": {
      "functionName": "invoke3",
      "handlerName": "index.invoke3",
    },
    "functionName": "invoke3",
    "functionTriggerMetadata": {},
    "functionTriggerName": "event",
    "handlerName": "helloHttpService.invoke3",
    "method": "invoke3",
    "middleware": [],
    "prefix": "/",
    "requestMetadata": [],
    "requestMethod": "",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "",
  },
  {
    "_category": 2,
    "_level": 0,
    "_paramString": "",
    "_pureRouter": "",
    "_weight": 0,
    "controllerId": "helloHttpService",
    "controllerMiddleware": [],
    "description": "",
    "funcHandlerName": "index.invoke4",
    "functionMetadata": {
      "functionName": "invoke4",
      "handlerName": "index.invoke4",
    },
    "functionName": "invoke4",
    "functionTriggerMetadata": {
      "handlerName": "index.invoke44",
    },
    "functionTriggerName": "event",
    "handlerName": "helloHttpService.invoke4",
    "method": "invoke4",
    "middleware": [],
    "prefix": "/",
    "requestMetadata": [],
    "requestMethod": "",
    "responseMetadata": [],
    "routerName": "",
    "summary": "",
    "url": "",
  },
]
`;
